### ISSUE TITLE
When typing a chat, the last letter doesn't appear

### ISSUE THREAD

**Issue by Kieaer** [Anuken/Mindustry#1943]
**Platform**: All desktop platforms (Windows/Linux/MacOS)

**Build**: Steam/release 104.10

**Issue**: When typing a chat, the last letter doesn't appear
![complex](https://user-images.githubusercontent.com/44261958/80611337-6f387880-8a75-11ea-96d2-746b3e84615e.gif)
Animated gif description
1. Typing ``Hello``
2. Typing ``안녕``
3. Typing ``안녕 ``&nbsp;2 times
3. Typing ``안녕`` again
5. Typing ``Hello``
6. Typing ``안녕`` and use arrow key

Korean structure
``ㅇ`` + ``ㅏ`` + ``ㄴ``= ``안``
``ㄴ`` + ``ㅕ`` + ``ㅇ`` = ``녕`` 

**Steps to reproduce**: 
1. Open chat popup
2. Just use asian chat (Japanese/Korean/Chinese)

**Link to mod(s) used, if applicable**: Vanilla

---

- [x] **I have searched the closed and open issues to make sure that this problem has not already been reported.**

---
This has been a problem since using TTF fonts.
I don't know where I'm getting keyboard input in ``TextField chatfreg``, and I know that there's only anuke to solve this.
The reason for posting this again is that it is inconvenient to communicate because the last letter is invisible, and sometimes the communication is wrong and the dispute occurs every day.

**Comment by Katsute** [comment]
This issue is caused by how Windows IME's work. When typing the strokes/steps Windows is actually typing it in its own input; when you complete a character Windows then inputs it to the Mindustry input. This is why when inputing a space the character appears; because Windows pushes the completed character to Mindustry before starting the next one.

Since Mindustry does not have its own IME implementation, the enter key sends the message instead of completing the currently typed character.

This could (I think) be fixed by ignoring the first enter press by the user if the last character in the input is an Asian one; this would allow the enter press to complete the character instead of sending the message without it completed.

**Comment by Katsute** [comment]
#1043

**Comment by Katsute** [comment]
I had a look at the java demo and the IME/IMF implementation seems successful: https://docs.oracle.com/javase/8/docs/technotes/guides/imf/index.html

**Comment by Kieaer** [comment]
Thanks. this may help me solve it myself.
maybe i should fix the arc engine.

It's hard to solve without using a bitmap font.

**Comment by Anuken** [comment]
What does this have to do with bitmap fonts?

**Comment by Kieaer** [comment]
@Anuken This problem didn't occur when using Bitmap Font (Remember 4.0 dev!). The reason seems to be explained above.

**Comment by Anuken** [comment]
To clarify, you were able to see every character you were typing in the alpha versions? That's completely unrelated. The type of font used would definitely not affect whether or not your characters are cut off in chat.

> the IME/IMF implementation seems successful:

That uses AWT, which is not compatible with Mindustry. Don't look at the standard Java library or UI toolkit.

**Comment by Anuken** [comment]
I use SDL, but it looks like GLFW has/had [a similar issue.](https://github.com/ocornut/imgui/issues/962)

**Comment by Anuken** [comment]
@Kieaer Can you show me what `안녕` being typed into Classic looks like?

**Comment by Kieaer** [comment]
@Anuken 3.5 version doesn't support

**Comment by Anuken** [comment]
How about [this version](https://github.com/Anuken/Mindustry/releases/tag/v63)?

**Comment by Kieaer** [comment]
![tag](https://user-images.githubusercontent.com/44261958/80767795-b6149400-8b83-11ea-8795-dc0cf6d57ae4.gif)
The last character is not visible in the input area, but the output is normal.

**Comment by Anuken** [comment]
It looks like the input system just doesn't handle enter properly when you're at the end of a word. I'll need to look into SDL's text edit events.

**Comment by Anuken** [comment]
@Kieaer What happens when you try to type the same characters into a text field like the player name, and press enter at the end? Is the character still missing?

**Comment by Katsute** [comment]
![ohno](https://user-images.githubusercontent.com/58778985/80871707-cb6ce980-8c7b-11ea-983c-e2b3aeed0f0d.gif)
In a message box pressing the enter key will complete typing the character, but will also create a new line at the same time. (because its the same key)

**Comment by Katsute** [comment]
![ohnoagain](https://user-images.githubusercontent.com/58778985/80871913-a036ca00-8c7c-11ea-9e1f-650a3553f8b2.gif)
Pressing enter the first time will complete the character without saving the schematic. (the code recieves the enter press before the textbox recieves the character)

Pressing enter for the second character will not complete the character but will save the schematic. (ths code sees a nonempty textbox before the character is inputted).

The issue:
1. Use types strokes via IME (code doesn't see this)
2. User presses enter
3. Mindustry sees the input is blank and does nothing
4. The IME inputs the character into the input
5. User types strokes via IME (code doesn't see this)
6. Use presses enter
7. Mindustry sees non-empty input and saves schematic
8. ⚠ IME wants to input text but can't because textbox no longer exists.

**Comment by Anuken** [comment]
Steps 7 and 8 should not happen in that order; if anything, step 7 should not happen at all. Any keys pressed during composition (including "enter" at the very end) shouldn't be visible to the application. Either SDL doesn't handle that case properly, or the my SDL backend doesn't process the events correctly.

**Comment by Kieaer** [comment]
> What happens when you try to type the same characters into a text field like the player name, and press enter at the end? Is the last letter still missing?

Yes.

**Comment by Anuken** [comment]
> Yes

So until you type another character, the IME output will never show up, regardless of whether enter is pressed? If so, that means @Katsute is incorrect and the problem is not related to order of input.

**Comment by Kieaer** [comment]
@Anuken Yes
![녹화_2020_05_03_12_12_45_493](https://user-images.githubusercontent.com/44261958/80897809-cff6d800-8d37-11ea-8786-9296c120b828.gif)
![ingame](https://user-images.githubusercontent.com/44261958/80897813-d5542280-8d37-11ea-8d54-42798e061347.gif)

**Comment by Kieaer** [comment]
Wait, I think I've seen this solve the same problem in other java games.
a.. minecraft!
I think the letters are complete in the input buffer, but the game engine doesn't recognize this.

**Comment by Anuken** [comment]
@Kieaer Show me what that looks like when you're inputting into the name text field in *join game*.

**Comment by Kieaer** [comment]
![녹화_2020_05_03_15_45_35_929](https://user-images.githubusercontent.com/44261958/80907852-32120600-8d55-11ea-8340-33f699419ad7.gif)

**Comment by ILiekMelons** [comment]
hmm

**Comment by Anuken** [comment]
@ILiekMelons Stop adding random unnecessary comments to threads; this is Github, not Discord. I get a notification every time you comment anywhere, and it's getting annoying. If it's not relevant to the discussion, don't say it.

**Comment by Remint32** [comment]
У меня происходит то же самое, только для случайных букв в чате

*Вот пример:*

![изображение](https://user-images.githubusercontent.com/13190430/95662232-1bf74580-0b3e-11eb-841a-254783302444.png)

![изображение](https://user-images.githubusercontent.com/13190430/95662235-26b1da80-0b3e-11eb-99fe-e5277b902a73.png)

Windows 10, Java 8 (JRE), BE-10361

Раньше такое **никогда** не происходило.

**Comment by Mr-Ao-Dragon** [comment]
Chinese, shou xin input, This bug also appears

**Comment by Anuken** [comment]
After testing this out thoroughly on a Windows VM with Korean text input, I figured out the exact sequence of events occurring here.

1. User composes something. I don't know anything about Korean, but in my tests pressing G and H made a combined character. The result isn't finished yet, so nothing is inputted into Mindustry.
2. User presses enter, *or any other key unrelated to composition.*
3. "Key Down" event is fired immediately, with the key *from step 2*. **This should not happen.** In a chat, this sends the message, missing the last character.
4. *1-2 frames later*: "Text Input" event is fired, with the text *from step 1*.
5. "Key Down" event is fired **again**, using key from step 2. This is indicated as a "repeat" event, *despite the fact that the key was only pressed once, with no time for repeats.* 
6. "Key Up" event is fired, once the key is released.

I've looked through my input handling code, and I can't spot any bugs anywhere - especially not ones that would make near-duplicate key events. A bug in SDL seem unlikely, but that's the only other explanation I can think of - *I should not be getting duplicate key events during composition*.

What's probably happening here is:
1. The SDL window gets the "enter pressed" event, and immediately sends it to my code, which causes the chat to close/send its message.
2. The IME gets the "enter pressed" event with a slight delay, and sends its current characters to SDL, along with the extra "enter" key.

This issue can't be worked around. The key events don't happen all in one frame, and they don't have any consistent spacing. Unless someone else can find out what I'm doing wrong or fix the supposed bug in SDL, this won't be solved.
